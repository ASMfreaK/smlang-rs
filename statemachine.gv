digraph G {
    rankdir="LR";
    node [fontname=Arial];
    s [shape=circle size=2 color="black" style=filled]
    
    s -> State1
	State2 [shape=box color="red" fillcolor="#ffbb33" style=filled]
	State3 [shape=box color="red" fillcolor="#ffbb33" style=filled]
	State1 [shape=box color="red" fillcolor="#ffbb33" style=filled]

	Event2 [label="Event2\n[_] / _"]
	Event1 [label="Event1\n[_] / _"]

	State2 -> Event2  -> State3;
	State1 -> Event1  -> State2;
}2 -> Event2  -> State3;
}ate1 -> Event1  -> State2;
	State3 -> Event3  -> State2;
}Event1  -> State2;
	State3 -> Event3  -> State1;
	State2 -> Event2  -> State3;
	State2 -> Event4  -> State1;
	State2 -> Event3  -> State1;
}