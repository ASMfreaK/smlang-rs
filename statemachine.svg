<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 2.40.1 (20161225.0304)
 -->
<!-- Title: G Pages: 1 -->
<svg width="677pt" height="62pt"
 viewBox="0.00 0.00 676.77 61.74" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 57.7401)">
<title>G</title>
<polygon fill="#ffffff" stroke="transparent" points="-4,4 -4,-57.7401 672.7716,-57.7401 672.7716,4 -4,4"/>
<!-- s -->
<g id="node1" class="node">
<title>s</title>
<ellipse fill="#000000" stroke="#000000" cx="18" cy="-26.8701" rx="18" ry="18"/>
<text text-anchor="middle" x="18" y="-23.1701" font-family="Arial" font-size="14.00" fill="#000000">s</text>
</g>
<!-- State1 -->
<g id="node2" class="node">
<title>State1</title>
<polygon fill="#ffbb33" stroke="#ff0000" points="131,-44.8701 72,-44.8701 72,-8.8701 131,-8.8701 131,-44.8701"/>
<text text-anchor="middle" x="101.5" y="-23.1701" font-family="Arial" font-size="14.00" fill="#000000">State1</text>
</g>
<!-- s&#45;&gt;State1 -->
<g id="edge1" class="edge">
<title>s&#45;&gt;State1</title>
<path fill="none" stroke="#000000" d="M36.1183,-26.8701C43.7154,-26.8701 52.8693,-26.8701 61.8896,-26.8701"/>
<polygon fill="#000000" stroke="#000000" points="61.927,-30.3702 71.927,-26.8701 61.9269,-23.3702 61.927,-30.3702"/>
</g>
<!-- Event1 -->
<g id="node5" class="node">
<title>Event1</title>
<ellipse fill="none" stroke="#000000" cx="246.196" cy="-26.8701" rx="79.3924" ry="26.7407"/>
<text text-anchor="middle" x="246.196" y="-30.6701" font-family="Arial" font-size="14.00" fill="#000000">Event1</text>
<text text-anchor="middle" x="246.196" y="-15.6701" font-family="Arial" font-size="14.00" fill="#000000">[guard] / action</text>
</g>
<!-- State1&#45;&gt;Event1 -->
<g id="edge2" class="edge">
<title>State1&#45;&gt;Event1</title>
<path fill="none" stroke="#000000" d="M131.1305,-26.8701C138.9026,-26.8701 147.6993,-26.8701 156.8824,-26.8701"/>
<polygon fill="#000000" stroke="#000000" points="156.9687,-30.3702 166.9686,-26.8701 156.9686,-23.3702 156.9687,-30.3702"/>
</g>
<!-- State2 -->
<g id="node3" class="node">
<title>State2</title>
<polygon fill="#ffbb33" stroke="#ff0000" points="420.3919,-44.8701 361.3919,-44.8701 361.3919,-8.8701 420.3919,-8.8701 420.3919,-44.8701"/>
<text text-anchor="middle" x="390.8919" y="-23.1701" font-family="Arial" font-size="14.00" fill="#000000">State2</text>
</g>
<!-- Event2 -->
<g id="node6" class="node">
<title>Event2</title>
<ellipse fill="none" stroke="#000000" cx="515.0818" cy="-26.8701" rx="58.8803" ry="26.7407"/>
<text text-anchor="middle" x="515.0818" y="-30.6701" font-family="Arial" font-size="14.00" fill="#000000">Event2</text>
<text text-anchor="middle" x="515.0818" y="-15.6701" font-family="Arial" font-size="14.00" fill="#000000">[_] / action</text>
</g>
<!-- State2&#45;&gt;Event2 -->
<g id="edge4" class="edge">
<title>State2&#45;&gt;Event2</title>
<path fill="none" stroke="#000000" d="M420.6385,-26.8701C428.4591,-26.8701 437.2294,-26.8701 446.1909,-26.8701"/>
<polygon fill="#000000" stroke="#000000" points="446.3495,-30.3702 456.3495,-26.8701 446.3494,-23.3702 446.3495,-30.3702"/>
</g>
<!-- State3 -->
<g id="node4" class="node">
<title>State3</title>
<polygon fill="#ffbb33" stroke="#ff0000" points="668.7716,-44.8701 609.7716,-44.8701 609.7716,-8.8701 668.7716,-8.8701 668.7716,-44.8701"/>
<text text-anchor="middle" x="639.2716" y="-23.1701" font-family="Arial" font-size="14.00" fill="#000000">State3</text>
</g>
<!-- Event1&#45;&gt;State2 -->
<g id="edge3" class="edge">
<title>Event1&#45;&gt;State2</title>
<path fill="none" stroke="#000000" d="M325.7399,-26.8701C334.5068,-26.8701 343.1313,-26.8701 351.0918,-26.8701"/>
<polygon fill="#000000" stroke="#000000" points="351.1301,-30.3702 361.1301,-26.8701 351.13,-23.3702 351.1301,-30.3702"/>
</g>
<!-- Event2&#45;&gt;State3 -->
<g id="edge5" class="edge">
<title>Event2&#45;&gt;State3</title>
<path fill="none" stroke="#000000" d="M573.9035,-26.8701C582.6317,-26.8701 591.4463,-26.8701 599.6551,-26.8701"/>
<polygon fill="#000000" stroke="#000000" points="599.6767,-30.3702 609.6767,-26.8701 599.6767,-23.3702 599.6767,-30.3702"/>
</g>
</g>
</svg>
