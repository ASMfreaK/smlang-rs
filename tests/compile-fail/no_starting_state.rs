extern crate smlang;

use smlang::statemachine;

statemachine! { 
    transitions {
        //~ ERROR No starting state defined, indicate the starting state with a *
        State1 + Event1 = State2,
        State2 + Event2 = State3,
    }
}

fn main() {}
