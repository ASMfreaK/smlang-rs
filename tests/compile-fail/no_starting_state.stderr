error: No starting state defined, indicate the starting state with a *.
  --> $DIR/no_starting_state.rs:5:1
   |
5  | / statemachine! {
6  | |     transitions: {
7  | |         //~ ERROR No starting state defined, indicate the starting state with a *
8  | |         State1 + Event1 = State2,
9  | |         State2 + Event2 = State3,
10 | |     }
11 | | }
   | |_^
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
